### PR TITLE
WIP: Lots of fixes from clippy

### PR DESCRIPTION
**Pull request by clarfonthey**
Will continue to add more changes to this until it's merged. This was the result of running cargo clippy on libcore and liballoc, although I didn't get close to fixing everything. I'll continue to add changes as I look through what clippy says, but I'm going to put this up now so that people can review the existing changes.

I'd be willing to split this into multiple commits if preferred.

### CODE REVIEW

**Comment by leonardo-m** [comment]
I think that changes to the Vec and similar parts should be done carefully, with benchmarks too.

**Comment by killercup** [comment]
This is awesome!

The `if … else { if … else … }` to `if … else if … else` changes make this diff larger than it needs to be. [Read it without whitespace changes](https://github.com/rust-lang/rust/pull/46960/files?w=1)

File: src/liballoc/str.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@
 //! We can explicitly specify `hello_world`'s lifetime as well:
 //!
 //! ```
-//! let hello_world: &'static str = "Hello, world!";
```
Comment (tinaun) [inline src/liballoc/str.rs]: this is in documentation showing how to explicitly specify a lifetime, so it shouldn't be removed

File: src/etc/dec2flt_table.py [hunk 2]
```diff
@@ -105,6 +105,7 @@ def error(f, e, z):
 
 //! Tables of approximations of powers of ten.
 //! DO NOT MODIFY: Generated by `src/etc/dec2flt_table.py`
+#![cfg_attr(feature = "cargo-clippy", allow(unreadable_literal))]
```
Comment (alexcrichton) [inline src/etc/dec2flt_table.py]: Can annotations like this be omitted? We don't run clippy in the main repository regularly so I'd prefer to not land annotations like this which are basically unused by all our infrastructure.

File: src/liballoc/clippy.toml [hunk 3]
```diff
@@ -0,0 +1 @@
+doc-valid-idents = ["TimSort"]
```
Comment (alexcrichton) [inline src/liballoc/clippy.toml]: Similar to above, can this be omitted?

File: src/liballoc/str.rs (hunk 1 above)
Comment (alexcrichton) [reply src/liballoc/str.rs]: Yes, can this be reverted?

File: src/liballoc/vec.rs [hunk 4]
```diff
@@ -2337,22 +2338,20 @@ impl<T> Iterator for IntoIter<T> {
         unsafe {
             if self.ptr as *const _ == self.end {
                 None
+            } else if mem::size_of::<T>() == 0 {
```
Comment (alexcrichton) [inline src/liballoc/vec.rs]: Can this be left as it was, it's a bit clearer in terms of control flow previously

File: src/liballoc/vec.rs [hunk 5]
```diff
@@ -2379,19 +2378,17 @@ impl<T> DoubleEndedIterator for IntoIter<T> {
         unsafe {
             if self.end == self.ptr {
                 None
-            } else {
-                if mem::size_of::<T>() == 0 {
-                    // See above for why 'ptr.offset' isn't used
-                    self.end = arith_offset(self.end as *const i8, -1) as *mut T;
+            } else if mem::size_of::<T>() == 0 {
```
Comment (alexcrichton) [inline src/liballoc/vec.rs]: Similar to above, can this be left as it was?

File: src/liballoc/vec.rs [hunk 6]
```diff
@@ -2485,7 +2482,7 @@ impl<'a, T> DoubleEndedIterator for Drain<'a, T> {
 impl<'a, T> Drop for Drain<'a, T> {
     fn drop(&mut self) {
         // exhaust self first
-        while let Some(_) = self.next() {}
+        for _ in self.by_ref() {}
```
Comment (alexcrichton) [inline src/liballoc/vec.rs]: Can this be left as it was? This is a pretty unsafe implementation so having it be as clear/simple as possible (less layers of abstraction) makes it easier to audit and see what's going on.

File: src/liballoc/vec_deque.rs [hunk 7]
```diff
@@ -2479,54 +2479,52 @@ impl<T> From<VecDeque<T>> for Vec<T> {
...
+                    ptr::copy(buf, buf.offset((cap - head) as isize), head);
+                    // shift the new thing to the start
+                    ptr::copy(buf.offset((tail - head) as isize), buf, len);
+                } else {
+                    // left hand block is the long one, we can do it in two!
+                    ptr::copy(buf, buf.offset((cap - tail) as isize), head);
+                    ptr::copy(buf.offset(tail as isize), buf, cap - tail);
+                }
             } else {
-                if (tail - head) >= cmp::min((cap - tail), head) {
```
Comment (alexcrichton) [inline src/liballoc/vec_deque.rs]: Similar to aove, can this be left as it was?

File: src/libcore/char_private.rs [hunk 8]
```diff
@@ -8,8 +8,8 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// NOTE: The following code was generated by "src/etc/char_private.py",
-//       do not edit directly!
+//! NOTE: The following code was generated by `src/etc/char_private.py`;
+//!       do not edit directly!
```
Comment (alexcrichton) [inline src/libcore/char_private.rs]: This I believe wasn't intended to be a doc comment, can the `!` be removed.

File: src/liballoc/str.rs (hunk 1 above)
Comment (clarcharr) [reply src/liballoc/str.rs]: Done

File: src/etc/dec2flt_table.py (hunk 2 above)
Comment (clarcharr) [reply src/etc/dec2flt_table.py]: Will do!

File: src/liballoc/clippy.toml (hunk 3 above)
Comment (clarcharr) [reply src/liballoc/clippy.toml]: Will do

File: src/libcore/char_private.rs (hunk 8 above)
Comment (clarcharr) [reply src/libcore/char_private.rs]: Okay!

File: src/liballoc/vec.rs (hunk 4 above)
Comment (clarcharr) [reply src/liballoc/vec.rs]: I can revert it but I'm a little confused how `else { if` is more clear than `else if`; it just adds unnecessary indentation IMHO.

**Comment by clarcharr** [comment]
I think it'll be best to just resubmit this PR as a series of smaller PRs for clarity. Closing this one.

**Pull request closed without merging**